src/java/org/apache/cassandra/io/sstable/indexsummary/IndexSummaryManager.java

```
@@ -99,24 +106,26 @@ private IndexSummaryManager(Supplier<List<T>> indexSummariesProvider)
         this.indexSummariesProvider = indexSummariesProvider;
 
         executor = executorFactory().scheduled(false, "IndexSummaryManager", Thread.MIN_PRIORITY);
-
-        long indexSummarySizeInMB = DatabaseDescriptor.getIndexSummaryCapacityInMiB();
-        int interval = DatabaseDescriptor.getIndexSummaryResizeIntervalInMinutes();
+        // TODO make format get from default format
+        format = DatabaseDescriptor.getSelectedSSTableFormat();
```

Because the format shouldn't always offer these options, we need to offer some way of recognizing an exposing them somewhat generally.

One way is to check here if the format implements an interface e.g. called `IndexSummaryConfiguration` which exposes these. BIG would implement it, and BTI would not. A further complication is that for this initialization we want to find the format that does use an index summary and get the options from there. If there's no such format, the manager should not be initialized at all (which is a good way of turning this and keycache off if not needed); if there are multiple, we should probably check that the options they provide match.

---

` if there are multiple, we should probably check that the options they provide match`
I think we can only get all the sstable formats provided from DatabaseDescriptor.getSSTableFormats() to fetch all the formats , there may exists some format that support IndexSummaryConfiguration but, I am not clear with `match` mean? And for me, `IndexSummaryManager` seems can not be table level, So I used `DatabaseDescriptor.getSelectedSSTableFormat()`.

---

The question is not about table level, but rather about data that already exists and is in other sstable formats (if you imagine a node with say 1 TiB of data that was written before the new format was available, and where we don't want to forcefully upgrade).

In other words, in a mixed-format environment, especially while we are switching from BIG to BTI. In this latter case the selected format (BTI) will not need index summaries and won't implement `IndexSummaryConfiguration` but we still need an `IndexSummaryManager` so that the old SSTables (in BIG format) can be dealt with properly.

---

thanks for your reply. In fact, what I want to focus more on is the matching issue of this format. Because if we have many configurable ones such as big-fast1/big-small1, they both support IndexSummaryConfiguration, so what should we do to make the action of match , which one should be picked up ?

---

I don't want to overthink it, especially as there is only one format that uses these at the moment -- if there are multiple and they don't match, we should raise an exception during the validation of the yaml.

---

Hi @blambov， long time no see, and I am back and continue working on this again. After  previous communication with you above , I already have some clear directions for modifying this patch . 

But I have a new question about table level configuration of [CASSANDRA-18534](https://issues.apache.org/jira/browse/CASSANDRA-18534) . 
the discussion in [CASSANDRA-18441](https://issues.apache.org/jira/browse/CASSANDRA-18441) has described that we will not support per-table configuration at this point , right ? so , that means we may only have two kinds of sstable format: big or bti. And there is not going to support the feature that contains more than two kinds of sstable format configuration  [here](https://github.com/apache/cassandra/blob/trunk/conf/cassandra.yaml#L1114), as I have said may be big/bti/big-small/bti-small and so on ? Am I right ?

---

We do not want to support per-table configuration at the moment. This means that we don't need to support format variations (e.g. bti-fast) either.

However, we should support formats that people can plug into the database by linking with a JAR that implements their custom format. We do not know what options such formats may use.